### PR TITLE
Replace QtWebKit with QtWebEngine

### PR DESCRIPTION
**Pull request by tkemmer**
Dear colleagues,
I have ported BALLView and both PresentaBALL and BALLaxy plugins to QtWebEngine (>= 5.5). Please have a look at the changes. So far, I have successfully tested the application on my Gentoo Linux machine using the BALL contrib but there are a few things that I'd like to sort out before merging…

Primarily, this boils down to the following two issues:
* There still seem to be major problems with QtWebEngine on systems using the open-source nouveau graphics driver for Nvidia cards:
 * https://bugreports.qt.io/browse/QTBUG-41242
 * https://bugreports.qt.io/browse/QTBUG-49243
 * https://bugs.freedesktop.org/show_bug.cgi?id=91632

 I briefly tested the new implementation on a second machine (Arch Linux 4.5.1, GeForce GTX 460, xf86-video-nouveau 1.0.12) and BALLView freezes or crashes immediately when enabling PresentaBALL or BALLaxy, showing a segfault caused by Chromium/nouveau in the logs:
 ```
[   68.416179] nouveau 0000:02:00.0: fifo: PBDMA0: 00040000 [] ch 5 [002fb18000 BALLView[751]] subc 0 mthd 0000 data 00000000
[   68.416194] nouveau 0000:02:00.0: gr: DATA_ERROR 00000004 [INVALID_VALUE] ch 5 [002fb18000 BALLView[751]] subc 0 class 9097 mthd 0ff0 data 20050004
[   68.416210] nouveau 0000:02:00.0: fifo: PBDMA0: 00040000 [] ch 5 [002fb18000 BALLView[751]] subc 0 mthd 0000 data 00000000
[   69.045176] Chrome_InProcGp[773]: segfault at 8 ip 00007f01dc94843c sp 00007f018b7fd110 error 4 in libdrm_nouveau.so.2.0.0[7f01dc944000+7000]
[  301.736514] nouveau 0000:02:00.0: fifo: PBDMA0: 00040000 [] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0000 data 00000000
[  301.736536] nouveau 0000:02:00.0: fifo: PBDMA0: 00040000 [] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0000 data 00000000
[  301.736554] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 000c data 00000000
[  301.736573] nouveau 0000:02:00.0: fifo: PBDMA0: 02000000 [] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0020 data 3f800000
[  301.736598] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0030 data 00000000
[  301.736617] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0034 data 20010674
[  301.736634] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0038 data 0000003f
[  301.736650] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 003c data 200140c5
[  301.736666] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0040 data 00000800
[  301.736682] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0044 data 20054088
[  301.736698] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0048 data 00000040
[  301.736714] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 004c data 00000800
[  301.736731] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0054 data 00000001
[  301.736747] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0058 data 00000000
[  301.736763] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 005c data 200240c3
[  301.736779] nouveau 0000:02:00.0: fifo: PBDMA0: 00200000 [ILLEGAL_MTHD] ch 5 [002fb18000 BALLView[840]] subc 0 mthd 0060 data 00000000
[  366.992148] nouveau 0000:02:00.0: BALLView[840]: failed to idle channel 5 [BALLView[840]]
[  381.991741] nouveau 0000:02:00.0: BALLView[840]: failed to idle channel 5 [BALLView[840]]
```
 Can anyone of you confirm this behavior? I haven't tried to compile qt with OpenGL ES yet, as is suggested in the bug ticket.

* I had quite some trouble to get QtWebEngineViews to work in our DockWidgets until globally setting `Qt::AA_DontCreateNativeWidgetSiblings` which disables the creation of native windows unless told otherwise (when docking a QtWebEngineView, WebEngine complained repeatedly about QOpenGLWidgets being used as native child widgets before freezing completely). Well, frankly, I didn't oversee all the implications of globally diabling native windows but I did expect at least *some* differences… yet I saw none. Do you see any problems here?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -286,13 +287,26 @@ IF (BALL_HAS_VIEW)
...
-                                       Qt5::Xml)
+	LIST(APPEND VIEW_DEP_LIBRARIES Qt5::Widgets
+	                               Qt5::OpenGL
+	                               Qt5::PrintSupport
+	                               Qt5::Network
+	                               Qt5::Test
+	                               Qt5::Xml)
+
+	IF(USE_QTWEBENGINE)
+		FIND_PACKAGE(Qt5WebEngineWidgets REQUIRED)
```
Comment (dstoeckel) [inline CMakeLists.txt]: I also had to add a `FIND_PACKAGE(Qt5WebEngine)` here. Also I think that REQUIRED should be removed, as we can build BALL without the WebEngine.

**Comment by dstoeckel** [comment]
With the modifications to the buildsystem above it builds and works nicely for me.

Unfortunately I cannot reproduce the crash with nouveau as I do not have Nvidia hardware that I can use for testing.

Adding `Qt::AA_DontCreateNativeWidgetSiblings` should be fine. This flag is basically the default on the Wayland platform, so it cannot hurt to already abolish native widgets.

File: CMakeLists.txt (hunk 1 above)
Comment (tkemmer) [inline CMakeLists.txt]: Hmm, shouldn't `USE_QTWEBENGINE` rather be disabled then? My line of thought was that, if one wants to `USE_QTWEBENGINE`, both `QtWebEngineWidgets` and `QtWebChannel` are mandatory dependencies. Just like:
```
IF(USE_CUDA)
	FIND_PACKAGE(CUDA REQUIRED)
ENDIF()
```

Ah, and I added you as contributors to my fork as I forgot to push the webengine branch to the main repo before creating the pull request.

**Comment by tkemmer** [comment]
After talking to Andreas about the nouveau problem we decided not to postpone the WegEngine migration any further and rather disable both PresentaBALL and BALLaxy plugins on the affected platforms for now.

If you don't have any objections, I will merge my request tomorrow.

**Comment by dstoeckel** [comment]
I am for merging, but I am curious how you want to disable the plugin on those systems. Are you planning to introduce a runtime check?

**Comment by tkemmer** [comment]
Exactly (see 9cbcbbb)! Fortunately, it seems that it is already sufficient to hide the WebEngine widgets to prevent crashes, so we simply replace them during runtime with a harmless error message in the respective dock widget without actually destroying them and still preserve the option to re-enable them through the settings dialog. Given the user is brave enough, that is. ;-)

**Comment by dstoeckel** [comment]
I like that! :-) Just one minor nitpick. In your last commit some cases of `type *ptr` slipped in. I think the policy is to use pointer to the left everywhere (e.g. `type* ptr`). I should probably add that to the coding style ...

**Comment by tkemmer** [comment]
Oops, you are right. The style should be (more) consistent now.

Also, I've removed the `GLRenderer` usage from `HTMLView`. Just to make sure: There was a problem with using `QOpenGLFunctions` rather than `gl.h`, right?

**Comment by philthiel** [comment]
Jepp, however, I'm not really deep into the View and especially rendering stuff. I assume that I initialized the QOpenGLFunctions at the wrong place.

**Pull request merged**